src/persist/src/storage.rs

```diff
@@ -211,6 +211,36 @@ pub trait Blob: BlobRead + Sized {
     async fn delete(&mut self, key: &str) -> Result<(), Error>;
 }
 
+/// An abstraction for a location in storage that supports an arbitrary blob of binary data
+/// and a sequence number, where the data and sequence number are conditionally
+/// updated.
+///
+/// Users are expected to use this API with consistently increasing sequence numbers
+/// to allow multiple processes across multiple machines to agree to a total order
+/// of the evolution of this binary data.
+#[async_trait]
+pub trait Consensus: Sized {
```

Comment:
Here's the latest version in my local branch fleshing out the initial skeleton of a real impl of new persist

```
/// WIP
#[derive(Debug)]
pub struct CompareAndSet {
    /// WIP
    pub current: SeqNo,
    /// WIP
    pub value: Option<Vec<u8>>,
}

/// WIP
/// - define and document the initial state
///
/// TODO: Finish fleshing this out with what we need to support incremental
/// state as well as bounded size (i.e. something like a `scan` and `truncate`).
#[async_trait]
pub trait Consensus: std::fmt::Debug {
    /// Returns a recently stored value.
    ///
    /// Because of concurrent writes, network delays, etc this is not guaranteed
    /// to be the most recently stored value.
    async fn peek(&self, deadline: Instant) -> Result<Option<(SeqNo, Vec<u8>)>, LocationError>;

    /// WIP
    async fn compare_and_set(
        &self,
        deadline: Instant,
        expected: SeqNo,
        new: SeqNo,
        data: Vec<u8>,
    ) -> Result<Result<(), CompareAndSet>, LocationError>;
}
```

differences for discussion:
- mine no longer has the `type Config` and `fn open` stuff so that the trait is object safe. this is (currently) necessary so that the handles don't have to be parameterized by the Blob and Consensus types. we didn't need this before because everything was pushed to a runtime through a channel and we won't need it again once we push the state machine stuff behind the `PersistShard` rpc, but it seems nice in the meantime? or, perhaps I'll figure out some other way to work around this, dunno yet. seems fine to start either way and we can adjust later
  - for the same reason, it also doesn't `: Sized` but does `: std::fmt::Debug` instead
- mine has LocationError which isn't checked in yet, so what you have is fine
- mine has deadlines. it's fine if the first real impls don't do anything with them but I want to make sure we get the plumbing right in all the uses of Consensus, so I'd like to add them here
- `peek` vs `head`. I don't really have a strong opinion here yet, seems fine to punt on the bikeshed for now
- `compare_and_set` vs `cas`. I have a medium preference for spelling this out
  - `expected` vs `current`. another bikeshed, I have a medium preference for `expected`
  - our return types are a bit different. I don't feel strongly here expect that it should return both the SeqNo and the value (not just SeqNo) when expected doesn't match

Reply:
Ack on all of these - I'm happy to make these changes

Reply:
I came up with a structure / way of writing about it that I really like. I defined a type `VersionedData` which is the same as your `CompareAndSet` type, but its just "data + a version number" and then consensus is just "an abstraction over a persistent [VersionedData] that can only be updated conditionally on version". I'll push the in-mem version asap!

Reply:
yeah, I saw this. I love it!